### PR TITLE
Use bundled provider packages and improve build / publish process

### LINKED ISSUE #1435: Use bundled packages and update CI to dual publish to DockerHub and Upbound Marketplace
### What problem are you facing?

Currently, `provider-aws` is only published to DockerHub, and is done so as two separate images:
- `crossplane/provider-aws`
- `crossplane/provider-aws-controller`

As of [Crossplane `v1.6.0`](https://github.com/crossplane/crossplane/releases/tag/v1.6.0), "bundled provider packages" are now supported, meaning that only one image needs to be published. [Crossplane `v1.7.0`](https://github.com/crossplane/crossplane/releases/tag/v1.7.0) allowed for these bundled packages to also be cached efficiently when using OCI descriptor annotations.

Additionally, the Upbound Marketplace now adds a number of additional features on top of traditional OCI registries, including:
- Validation of Crossplane packages.
- Automatic documentation generation for all types in a package.
- Additional xpkg extension capabilities, such as attaching examples to API types.

An example of how this looks for `provider-aws@v0.24.1` can be found here: https://marketplace.upbound.io/providers/crossplane/provider-aws/v0.24.1

### How could Crossplane help solve your problem?

We should update `provider-aws` to use bundled packages and publish to both DockerHub and the Upbound Marketplace.

### PR DESCRIPTION
**Pull request by hasheddan**
### Description of your changes

Drops all deprecated annotations from the provider-aws crossplane.yaml.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

Drops the spec from the crossplane.yaml as bundled images will now be
used for provider-aws.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

Renames the single image that is being built as base to provider-aws.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

Updates package base image to use distroless static nonroot instead of
defaulting to outdated alpine image.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

Updates provider controller image building to use imagelight for
enhanced flexibility. Publish and promotion is deferred to xpkg
machinery.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

Adds the xpkg machinery to Makefile and updates other components to be
compatible.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

Fixes #1435 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

I tested this by overriding the registries to `xpkg.upbound.io/dan` and `index.docker.io/hasheddan`:

Build:
```
🤖 (stack-aws) VERSION=v0.32.0 make build.all
09:12:40 [ .. ] verify go modules dependencies have expected content
all modules verified
09:12:43 [ OK ] go modules dependencies verified
09:12:43 [ .. ] go build linux_amd64
github.com/crossplane-contrib/provider-aws/cmd/provider
09:12:56 [ OK ] go build linux_amd64
09:12:56 [ .. ] go build linux_arm64
github.com/crossplane-contrib/provider-aws/cmd/provider
09:13:06 [ OK ] go build linux_arm64
09:13:06 [ .. ] docker build build-b3b12a66/provider-aws-amd64
[+] Building 11.6s (8/8) FINISHED                                                                                                                          
 => [internal] load .dockerignore                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                       0.0s
 => [internal] load build definition from Dockerfile                                                                                                  0.0s
 => => transferring dockerfile: 301B                                                                                                                  0.0s
 => [internal] load metadata for gcr.io/distroless/static@sha256:1f580b0a1922c3e54ae15b0758b5747b260bd99d39d40c2edb3e7f6e2452298b                     0.5s
 => CACHED [1/2] FROM gcr.io/distroless/static@sha256:1f580b0a1922c3e54ae15b0758b5747b260bd99d39d40c2edb3e7f6e2452298b                                0.0s
 => => resolve gcr.io/distroless/static@sha256:1f580b0a1922c3e54ae15b0758b5747b260bd99d39d40c2edb3e7f6e2452298b                                       0.0s
 => [internal] load build context                                                                                                                     1.5s
 => => transferring context: 148.01MB                                                                                                                 1.5s
 => [2/2] ADD bin/linux_amd64/provider /usr/local/bin/crossplane-aws-provider                                                                         1.3s
 => exporting to oci image format                                                                                                                     8.2s
 => => exporting layers                                                                                                                               6.7s
 => => exporting manifest sha256:24005ad7452175b4fa031d99f970ef7b4ffd6cbc8d3d9930db67a468a466033e                                                     0.0s
 => => exporting config sha256:ce8ff5aee677bfb7a84d6461c9bb258ce3d5ede94a63cf512088ed6ae57fec09                                                       0.0s
 => => sending tarball                                                                                                                                1.4s
 => importing to docker                                                                                                                               1.0s
09:13:19 [ OK ] docker build build-b3b12a66/provider-aws-amd64
09:13:19 [ .. ] Building package provider-aws-v0.32.0.xpkg for linux_amd64
xpkg saved to /home/dan/code/github.com/crossplane/stack-aws/_output/xpkg/linux_amd64/provider-aws-v0.32.0.xpkg
09:13:29 [ OK ] Built package provider-aws-v0.32.0.xpkg for linux_amd64
09:13:29 [ .. ] docker build build-b3b12a66/provider-aws-arm64
[+] Building 10.7s (8/8) FINISHED                                                                                                                          
 => [internal] load build definition from Dockerfile                                                                                                  0.0s
 => => transferring dockerfile: 301B                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                       0.0s
 => [internal] load metadata for gcr.io/distroless/static@sha256:1f580b0a1922c3e54ae15b0758b5747b260bd99d39d40c2edb3e7f6e2452298b                     0.9s
 => [internal] load build context                                                                                                                     2.5s
 => => transferring context: 139.69MB                                                                                                                 2.5s
 => CACHED [1/2] FROM gcr.io/distroless/static@sha256:1f580b0a1922c3e54ae15b0758b5747b260bd99d39d40c2edb3e7f6e2452298b                                0.0s
 => => resolve gcr.io/distroless/static@sha256:1f580b0a1922c3e54ae15b0758b5747b260bd99d39d40c2edb3e7f6e2452298b                                       0.0s
 => [2/2] ADD bin/linux_arm64/provider /usr/local/bin/crossplane-aws-provider                                                                         0.6s
 => exporting to oci image format                                                                                                                     6.7s
 => => exporting layers                                                                                                                               5.4s
 => => exporting manifest sha256:732aef428a096e7510ce84bfc4e29b37b8d2249ba925246fad6558b680614fa5                                                     0.0s
 => => exporting config sha256:e8776253bb5ae33e3a3e0e44b7d127487b8912276923bf1efc5e80f6f6ad5f71                                                       0.0s
 => => sending tarball                                                                                                                                1.3s
 => importing to docker                                                                                                                               0.9s
09:13:40 [ OK ] docker build build-b3b12a66/provider-aws-arm64
09:13:40 [ .. ] Building package provider-aws-v0.32.0.xpkg for linux_arm64
xpkg saved to /home/dan/code/github.com/crossplane/stack-aws/_output/xpkg/linux_arm64/provider-aws-v0.32.0.xpkg
09:13:50 [ OK ] Built package provider-aws-v0.32.0.xpkg for linux_arm64
```

Publish:
```
🤖 (stack-aws) VERSION=v0.32.0 BRANCH_NAME=master make -j2 publish
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
09:14:48 [ .. ] Pushing package xpkg.upbound.io/dan/provider-aws:v0.32.0
09:14:48 [ .. ] Skipping image publish for docker.io/provider-aws:v0.32.0
Publish is deferred to xpkg machinery
09:14:48 [ OK ] Image publish skipped for docker.io/provider-aws:v0.32.0
09:14:48 [ .. ] Pushing package index.docker.io/hasheddan/provider-aws:v0.32.0
xpkg pushed to index.docker.io/hasheddan/provider-aws:v0.32.0
09:15:36 [ OK ] Pushed package index.docker.io/hasheddan/provider-aws:v0.32.0
xpkg pushed to xpkg.upbound.io/dan/provider-aws:v0.32.0
09:15:50 [ OK ] Pushed package xpkg.upbound.io/dan/provider-aws:v0.32.0
```

Promote:
```
🤖 (stack-aws) VERSION=v0.32.0 make -j2 promote
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
09:16:55 [ .. ] Promoting package from index.docker.io/hasheddan/provider-aws:v0.32.0 to index.docker.io/hasheddan/provider-aws:master
09:16:55 [ .. ] Skipping image promotion from docker.io/provider-aws:v0.32.0 to docker.io/provider-aws:master
Promote is deferred to xpkg machinery
09:16:55 [ OK ] Image promotion skipped for docker.io/provider-aws:v0.32.0 to docker.io/provider-aws:master
docker.io/hasheddan/provider-aws:master
09:16:56 [ OK ] Promoted package from index.docker.io/hasheddan/provider-aws:v0.32.0 to index.docker.io/hasheddan/provider-aws:master
🤖 (stack-aws) VERSION=v0.32.0 CHANNEL=stable make -j2 promote
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
09:17:15 [ .. ] Promoting package from index.docker.io/hasheddan/provider-aws:v0.32.0 to index.docker.io/hasheddan/provider-aws:stable
09:17:15 [ .. ] Skipping image promotion from docker.io/provider-aws:v0.32.0 to docker.io/provider-aws:stable
Promote is deferred to xpkg machinery
09:17:15 [ OK ] Image promotion skipped for docker.io/provider-aws:v0.32.0 to docker.io/provider-aws:stable
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
09:17:15 [ .. ] Skipping image promotion from docker.io/provider-aws:v0.32.0 to docker.io/provider-aws:v0.32.0-stable
Promote is deferred to xpkg machinery
09:17:15 [ OK ] Image promotion skipped for docker.io/provider-aws:v0.32.0 to docker.io/provider-aws:v0.32.0-stable
docker.io/hasheddan/provider-aws:stable
docker.io/hasheddan/provider-aws:v0.32.0-stable
09:17:17 [ OK ] Promoted package from index.docker.io/hasheddan/provider-aws:v0.32.0 to index.docker.io/hasheddan/provider-aws:stable
```

Examine package contents from dockerhub: https://explore.ggcr.dev/?image=hasheddan%2Fprovider-aws%3Av0.32.0

Package (private for now) in Upbound Marketplace:
Screenshot from 2022-08-29 22-15-51

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,4 @@
 [submodule "build"]
 	path = build
-	url = https://github.com/upbound/build
+	url = https://github.com/hasheddan/build
+	branch = xpkg-build
```
Comment (hasheddan) [inline .gitmodules:4]: Will be dropped after https://github.com/upbound/build/pull/203 is merged.

**Review by hasheddan — COMMENTED** [review]
Looks like `build` updates might go ahead and force us onto Go 1.18 here.

**Comment by hasheddan** [comment]
go 1.18 changes in flight in https://github.com/crossplane-contrib/provider-aws/pull/1410

**Comment by chlunde** [comment]
Looks good to me, but I don't know exactly what to look for. Looks like you have tested it thoroughly. One thing I would still like to be easily possible is to build and push images to other container registries (such as GHCR) to test PRs/branches/forks, I assume this is still possible by adapting the workflow a bit.

**Review by chlunde — APPROVED** [review]

**Comment by hasheddan** [comment]
> One thing I would still like to be easily possible is to build and push images to other container registries (such as GHCR) to test PRs/branches/forks, I assume this is still possible by adapting the workflow a bit.

@chlunde absolutely 👍🏻 in fact you should be able to just do `make build && XPKG_REG_ORG='your-registry.com/your-org' make publish` if you want to use the existing machinery. You can also use `up` directly to push the packages in `_output/xpkg/{OS}_{ARCH}/` or `docker load` from that directory and push with `docker`. If you find yourself not being able to execute your ideal workflow we swing back around and adjust the machinery.

**Pull request merged**